### PR TITLE
nixos/snapper: add snapper opts

### PR DESCRIPTION
**Pull request by Djabx**
## Description of changes

Add some options to the configuration of snapper:
Cf: https://github.com/openSUSE/snapper/blob/master/data/default-config

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Djabx** [comment]
Result of `nixpkgs-review pr 321237` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

File: nixos/modules/services/misc/snapper.nix [hunk 1]
```diff
@@ -78,6 +78,54 @@ let
         Defines whether hourly snapshots should be created.
       '';
     };
+
+    TIMELINE_LIMIT_HOURLY = mkOption {
```
Comment (superherointj) [inline nixos/modules/services/misc/snapper.nix:82]: The format for NixOS modules is: `timelineLimitHourly`

Comment (Djabx) [reply nixos/modules/services/misc/snapper.nix:82]: For what I understand, options are copied "as is" in the configuration file.

```
        // (mapAttrs'
          (name: subvolume: nameValuePair "snapper/configs/${name}" ({
            text = lib.generators.toKeyValue { inherit mkKeyValue; } (filterAttrs (k: v: v != defaultOf k) subvolume);
          }))
          cfg.configs)
```

End I'm not sur I want to change this magic line.

Comment (superherointj) [reply nixos/modules/services/misc/snapper.nix:82]: Right. Nevermind.

**Comment by superherointj** [comment]
Can you confirm which formatter you have used? It doesn't match mine here ([nixfmt-rfc-style](https://search.nixos.org/packages?query=nixfmt), from unstable).
If you drop the formatting I can merge this.
Or you need to fix the formatting. And do it last.

**Comment by Djabx** [comment]
I'm using: https://marketplace.visualstudio.com/items?itemName=jnoortheen.nix-ide 

But will revert for nixfmt-rfc-style

**Comment by superherointj** [comment]
> I'm using: https://marketplace.visualstudio.com/items?itemName=jnoortheen.nix-ide

IMO this extension should have `nixfmt` replaced with `nixfmt-rfc-style`.

> But will revert for nixfmt-rfc-style

`nixfmt-rfc-style` is currently the official one.

Code-wise your PR is good.

Still, commit messages are ending with a `.` and that's not usual.

From: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

> For consistency, there should not be a period at the end of the commit message's summary line (the first line of the commit message).

**Comment by Djabx** [comment]
@superherointj Re-format with `nixfmt-rfc-style` and amend commit.

**Review by superherointj — APPROVED** [review]

**Comment by Djabx** [comment]
Thanks @superherointj for all the hints !

**Comment by piegamesde** [comment]
Why was `types.str` chosen for configuration values which are clearly meant to be integers?

Moreover, adding new options into a freeform type is a breaking change. This should not have been backported.

**Comment by Djabx** [comment]
> Why was `types.str` chosen for configuration values which are clearly meant to be integers?

Agree on that, but the config file require quotes on numbers: http://snapper.io/manpages/snapper-configs.html

**Comment by Djabx** [comment]
> Moreover, adding new options into a freeform type is a breaking change. This should not have been backported.

How is it a breaking change ?

**Comment by piegamesde** [comment]
> Agree on that, but the config file require quotes on numbers

It doesn't, they just quote the example values in the documentation. (Even if they did, we should still allow numeric values and convert them in `mkValue` for convenience.)

> How is it a breaking change ?

By introducing a type check, you are restricting the set of possible allowed values. So if somebody has for example `TIMELINE_LIMIT_HOURLY = 5;` in their config this used to work but won't anymore.

**Comment by superherointj** [comment]
@Djabx Would you make the asked changes?

**Comment by Djabx** [comment]
> > How is it a breaking change ?
> 
> By introducing a type check, you are restricting the set of possible allowed values. So if somebody has for example `TIMELINE_LIMIT_HOURLY = 5;` in their config this used to work but won't anymore.

So your scenario is, somebody may have right some values in their configuration nix file, that was completly useless, never written in their /etc/snapper/configs/<file> and it could create a breaking change ?

**Comment by superherointj** [comment]
> > @Djabx Would you make the asked changes?
> 
> I could... 

Thanks for understanding. And I'm sorry for this unfortunate regression event.

> but I'm not sure what changes ?

Sure, we should clarify what should be done.

> Should we revert from the release-24.05 ? 

Since `piegamesde` requested the revert. It is reverted now.
Let's fix whatever is needed and move forward again.

> Allow int values ? Require int values ?

My guess is, `piegamesde` wants it to be int value, as required, not optionally.

> Should I change generate a config file with quotes ?

Seems so.

> Should we move to your suggestion @superherointj (cf: https://github.com/NixOS/nixpkgs/pull/321237/files#r1647653048 ) ? I think it would be a greatter breaking change, but maybe for a better good.

Please disregard my initial suggestion.

**Comment by superherointj** [comment]
@Djabx Would you mind adding yourself as a maintainer of this nixos module?

**Comment by superherointj** [comment]
> So your scenario is, somebody may have right some values in their configuration nix file, that was completly useless, never written in their /etc/snapper/configs/ and for you it's a breaking change ? That mean that adding any parameter is a breaking change ?

The realm of pragmatism (what is likely to happen) and the possibilities (what can happen) can't reconcile.
Let's agree to disagree on this. And let's just move forward, please.

It's just unfortunate, for everyone here, I know.

Thanks for the understanding of everyone.

**Comment by Djabx** [comment]
> @Djabx Would you be interested in being a maintainer of this NixOS module?

Euh... I'm just a snapper user, interested in time backup and struggling with lot of backup not cleanup.

I'm not a nix expert as you may have guess :)

**Comment by superherointj** [comment]
> I'm not a nix expert as you may have guess :)

That's a species in extinction, don't let that hold you back.
In practice, you would just be notified of changes to the module.
You can decide to do nothing. Or input something that you know to be right/true.
In case you get something wrong, you learn something new and correct your understandings.
We are all learning.

**Pull request merged**